src/envoy/http/authn/http_filter.cc

```diff
@@ -30,59 +57,211 @@ void AuthenticationFilter::onDestroy() {
-        std::string user, ip;
-        int port = 0;
-        bool ret = false;
-        ret = mtls_authn.GetSourceUser(&user);
-        if (ret) {
-          ENVOY_LOG(debug, "AuthenticationFilter: the source user is {}", user);
-        } else {
-          ENVOY_LOG(error,
-                    "AuthenticationFilter: GetSourceUser() returns false!");
-        }
-        ret = mtls_authn.GetSourceIpPort(&ip, &port);
-        if (ret) {
-          ENVOY_LOG(debug,
-                    "AuthenticationFilter: the source ip is {}, the source "
-                    "port is {}",
-                    user, port);
-        } else {
-          ENVOY_LOG(error,
-                    "AuthenticationFilter: GetSourceIpPort() returns false!");
-        }
+        ENVOY_LOG(warn,
+                  "Source authentiation payload doesn't contain x509 nor jwt "
+                  "payload.");
       }
+    }
+
+    const auto& rule =
+        GetCredentialRuleOrDefault(context_.peer_user(), config_);
+    if (rule.origins_size() == 0) {
+      switch (rule.binding()) {
+        case iaapi::CredentialRule::USE_ORIGIN:
+          // Validation should reject policy that have rule to USE_ORIGIN but
+          // does not provide any origin method so this code should
+          // never reach. However, it's ok to treat it as authentication
+          // fails.
+          ENVOY_LOG(
+              warn,
+              "Principal is binded to origin, but not methods specified in "
+              "rule {}",
+              rule.DebugString());
+          onAuthenticateOriginDone(headers, &rule, 0, nullptr,
+                                   IstioAuthN::Status::FAILED);
+          break;
+        case iaapi::CredentialRule::USE_PEER:
+          // On the other hand, it's ok to have no (origin) methods if
+          // rule USE_SOURCE
+          onAuthenticateOriginDone(headers, &rule, 0, nullptr,
+                                   IstioAuthN::Status::SUCCESS);
+          break;
+        default:
+          // Should never come here.
+          ENVOY_LOG(error, "Invalid binding value for rule {}",
+                    rule.DebugString());
+          break;
+      }
+      return;
+    }
+    authenticateOrigin(
+        *headers, rule.origins(0),
+        std::bind(&AuthenticationFilter::onAuthenticateOriginDone, this,
+                  headers, &rule, 0, std::placeholders::_1,
+                  std::placeholders::_2));
+  }
+}
+
+void AuthenticationFilter::authenticateOrigin(
+    HeaderMap& headers, const iaapi::OriginAuthenticationMethod& method,
+    const AuthenticateDoneCallback& done_callback) {
+  ValidateJwt(headers, method.jwt(), done_callback);
+}
+
+void AuthenticationFilter::onAuthenticateOriginDone(
+    HeaderMap* headers, const iaapi::CredentialRule* rule, int method_index,
+    std::unique_ptr<IstioAuthN::AuthenticatePayload> payload,
+    const IstioAuthN::Status& status) {
+  if (status == IstioAuthN::Status::FAILED) {
```

Comment:
For Status &status, only its value is used here. It will be safer for the Status &status be passed by value, not passed by reference.

Reply:
done